In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.__version__

'3.0.6'

In [3]:
car_data = pd.read_csv("car_fuel_efficiency_2026.csv")

In [4]:
car_data.tail()

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
9995,2005,USA,Gasoline,All-wheel drive,4,2470,6,271.0,4690,18.9,27.4
9996,1993,Asia,Gasoline,All-wheel drive,2,2300,6,244.0,4370,19.0,30.0
9997,2014,USA,Gasoline,All-wheel drive,3,2480,6,267.0,4590,18.4,28.1
9998,2004,USA,Hybrid,Rear-wheel drive,5,2180,6,270.0,4450,20.9,32.2
9999,2011,USA,Diesel,Front-wheel drive,4,2380,6,295.0,4370,17.6,30.2


In [5]:
car_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   model_year           10000 non-null  int64  
 1   origin               10000 non-null  str    
 2   fuel_type            10000 non-null  str    
 3   drivetrain           10000 non-null  str    
 4   num_doors            10000 non-null  int64  
 5   engine_displacement  10000 non-null  int64  
 6   num_cylinders        10000 non-null  int64  
 7   horsepower           9123 non-null   float64
 8   vehicle_weight       10000 non-null  int64  
 9   acceleration         9736 non-null   float64
 10  fuel_efficiency_mpg  10000 non-null  float64
dtypes: float64(3), int64(5), str(3)
memory usage: 859.5 KB


In [6]:
car_data.describe()

,model_year,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
count,10000.000000,10000.00000,10000.000000,10000.000000,9123.000000,10000.000000,9736.000000,10000.000000
mean,1999.549500,3.50400,2268.807000,6.035200,254.446016,4286.754000,18.888044,29.997700
std,14.207925,0.95209,183.447355,0.542578,22.844613,266.212491,1.279543,2.930245
min,1975.000000,2.00000,1590.000000,4.000000,171.000000,3320.000000,14.000000,19.800000
25%,1987.000000,3.00000,2150.000000,6.000000,239.000000,4110.000000,18.000000,28.000000
50%,2000.000000,4.00000,2270.000000,6.000000,254.000000,4280.000000,18.900000,30.000000
75%,2012.000000,4.00000,2390.000000,6.000000,270.000000,4470.000000,19.700000,31.900000
max,2024.000000,5.00000,3110.000000,8.000000,334.000000,5460.000000,23.800000,41.200000


In [7]:
car_data.index

RangeIndex(start=0, stop=10000, step=1)

In [8]:
car_data.columns

Index(['model_year', 'origin', 'fuel_type', 'drivetrain', 'num_doors',
       'engine_displacement', 'num_cylinders', 'horsepower', 'vehicle_weight',
       'acceleration', 'fuel_efficiency_mpg'],
      dtype='str')

In [9]:
set(car_data["fuel_type"])

{'Diesel', 'Gasoline', 'Hybrid'}

In [10]:
car_data[
    (car_data["origin"] == "Asia") &
    (car_data["fuel_efficiency_mpg"] == car_data["fuel_efficiency_mpg"].max())
]

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
6354,2022,Asia,Diesel,Front-wheel drive,4,1990,6,225.0,3870,17.2,41.2


In [11]:
car_data["horsepower"]

0       243.0
1       272.0
2       267.0
3       258.0
4       304.0
        ...  
9995    271.0
9996    244.0
9997    267.0
9998    270.0
9999    295.0
Name: horsepower, Length: 10000, dtype: float64

In [12]:
horsepower_median = car_data["horsepower"].median()
horsepower_median

np.float64(254.0)

In [13]:
horsepower_mode = car_data["horsepower"].mode()[0]
horsepower_mode

np.float64(252.0)

In [14]:
car_data["horsepower"] = car_data["horsepower"].fillna(horsepower_mode)

In [15]:
horsepower_median_filled = car_data["horsepower"].median()
horsepower_median_filled

np.float64(252.0)

car_data["origin"] == "Asia" gives a Series of True/False values, one per row. Using it inside car_data[...] keeps only the rows where it is True, so asia_cars holds only Asian-origin cars. Their original row labels are kept, so the labels won't be 0, 1, 2…
asia_cars[["vehicle_weight", "model_year"]] picks two columns. The double brackets matter: a list of column names returns a DataFrame, while a single name like ["vehicle_weight"] would return a Series.
.head(7) takes the first 7 rows in the order they appear in the CSV. This is not a random sample, so the result depends on the file's row order.
.values turns the DataFrame into a plain NumPy array (first_7.to_numpy() is the newer spelling). The labels are dropped and you get a 7×2 integer matrix:

      weight  year
    [[4240, 1996],
     [4310, 1992],
     ...
     [4450, 2019]]
Each row is one observation (a car) and each column is one feature. In ML notation this is X, with shape (n_samples, n_features) = (7, 2).


In [16]:
asia_cars = car_data[car_data["origin"] == "Asia"]
subset = asia_cars[["vehicle_weight", "model_year"]]
first_7 = subset.head(7)
X = first_7.values
X

array([[4240, 1996],
       [4310, 1992],
       [4230, 1997],
       [4420, 1978],
       [3950, 2007],
       [4750, 2021],
       [4450, 2019]])

XTX = X.T @ X
X.T is the transpose, shape (2, 7): features become rows.
@ is matrix multiplication (the same as np.matmul). A (2×7) times a (7×2) gives a 2×2 matrix.
Each entry of the result is a dot product between two feature columns:

weight	year
weight	Σ weight² = 131,950,500	Σ weight·year = 60,750,580
year	Σ year·weight = 60,750,580	Σ year² = 28,041,424
The matrix is symmetric because the off-diagonal entries are the same sum. It is always square (n_features × n_features), no matter how many rows X has. That is why it can be inverted even though X itself (7×2) can't.


In [17]:
XTX = X.T @ X
XTX

array([[131950500,  60750580],
       [ 60750580,  28041424]])

XTX_inv = np.linalg.inv(XTX)
This finds the matrix (XᵀX)⁻¹ such that XᵀX @ XTX_inv ≈ I, the identity matrix. You can check this yourself with XTX @ XTX_inv; you should get roughly [[1, 0], [0, 1]], with tiny floating-point errors off the diagonal.

The entries are tiny (~1e-6) because the entries of XᵀX are huge (~1e8). Inverting a matrix roughly swaps big scale for small scale.

The inverse exists only if the columns of X are linearly independent, meaning neither column is a multiple of the other. Weight and year are not proportional, so it works. If you had duplicate features, inv would raise a LinAlgError ("singular matrix") or return huge, unstable numbers.


In [18]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 2.96830537e-06, -6.43071025e-06],
       [-6.43071025e-06,  1.39675281e-05]])

y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
This is the target vector: one value per row of X, so its length is 7. The homework supplies these numbers directly; they don't come from the dataset. Row i of X is paired with y[i].


In [19]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

w = XTX_inv @ X.T @ y
This is the closed-form solution to linear least squares:

$$
w = (X^\top X)^{-1} X^\top y
$$

What it solves: find w = [w₁, w₂] so that the predictions


ŷ = X @ w  →  ŷᵢ = w₁·weightᵢ + w₂·yearᵢ
are as close as possible to y, measured by the sum of squared errors ‖Xw − y‖².

Where the formula comes from:

You can't solve Xw = y exactly: there are 7 equations and only 2 unknowns, so the system is overdetermined.
Take the gradient of ‖Xw − y‖² with respect to w and set it to zero. That gives XᵀX w = Xᵀy, the "normal equations".
Multiply both sides by (XᵀX)⁻¹ to get the formula above.
The shapes, evaluated left to right:

XTX_inv @ X.T → (2×2)(2×7) = (2×7)
… @ y → (2×7)(7,) = (2,)
Result: w ≈ [0.1364, 0.2327], where

w₁ = 0.1364 is the weight on vehicle_weight
w₂ = 0.2327 is the weight on model_year
As a sanity check, for the first car: 0.1364·4240 + 0.2327·1996 ≈ 578 + 465 ≈ 1043, which is close to its target of 1100.

Note: there's no intercept (bias) term. X has no column of 1s, so the model is forced through the origin: ŷ = w₁x₁ + w₂x₂, with no + w₀. That's what the homework asks for. In the course's later lessons you'll add np.ones as a first column so the model can learn w₀.


In [20]:
w = XTX_inv @ X.T @ y
w

array([0.13644777, 0.2327492 ])

w_sum = w.sum()   # 0.36919696904925486
This adds the two weights, 0.13645 + 0.23275 ≈ 0.3692. The number has no statistical meaning on its own. It's a single value the homework uses to check your answer.


In [21]:
w_sum = w.sum()
w_sum

np.float64(0.36919696904925486)

Numerical stability: computing an explicit inverse is fine at this size. In practice, np.linalg.solve(XTX, X.T @ y) or np.linalg.lstsq(X, y, rcond=None)[0] give the same result more accurately and are preferred, especially with badly scaled features like these (weight ≈ 4000, year ≈ 2000).
The earlier fillna in cell [20] doesn't affect this part, because horsepower isn't one of the features used here.

#Why is it necesssary to transpose the matrix in cell 23?

The transpose is there because the normal equation needs the 2×2 feature-by-feature matrix XᵀX, and X.T @ X is the only way to get it from a 7×2 matrix. Without .T, you either can't multiply at all or you get the wrong matrix.

1. The shapes don't line up without it
In matrix multiplication, the inner dimensions must match: (a×b) @ (b×c) → (a×c). X has shape (7, 2), with 7 cars and 2 features.

Expression	Shapes	Result
X @ X	(7×2) @ (7×2)	❌ Error, because 2 ≠ 7
X @ X.T	(7×2) @ (2×7)	7×7 matrix, the wrong thing (see section 3)
X.T @ X	(2×7) @ (7×2)	✅ 2×2 matrix, what we want
2. The math asks for exactly XᵀX
We want the w that minimizes the squared error ‖Xw − y‖². Setting its gradient to zero gives:

$$
\nabla_w |Xw - y|^2 = 2X^\top(Xw - y) = 0 ;;\Rightarrow;; X^\top X, w = X^\top y
$$

The Xᵀ comes straight out of the calculus. Its job is to collapse the 7 per-car errors into one equation per feature, giving 2 equations for the 2 unknowns in w. The system Xw = y itself has 7 equations and 2 unknowns, which usually has no exact solution. Multiplying by Xᵀ turns it into a solvable 2×2 system.

3. What X.T @ X actually computes
When you transpose X, its columns (features) become rows. So each entry of X.T @ X is a dot product between two feature columns, summed across all 7 cars:


               X.T (2×7)                          X (7×2)
weight → [4240 4310 4230 4420 3950 4750 4450]     [[4240, 1996],
year   → [1996 1992 1997 1978 2007 2021 2019]  @   [4310, 1992],
                                                   ...        ]
Top-left: weight · weight = Σ weight²
Off-diagonal: weight · year
Bottom-right: year · year
The result describes how the features relate to each other, which is what the solver needs.

X @ X.T would instead take dot products between rows, meaning cars. That gives a 7×7 matrix of how similar each car is to every other car. It has the wrong shape for w. Also, because it is built from only 2 features, its rank is at most 2, so it's singular and np.linalg.inv would fail or return garbage.

4. Why a square matrix matters
Only square matrices can have an inverse. X is 7×2, so X⁻¹ doesn't exist. XᵀX is always square (n_features × n_features), however many rows you have. It is also invertible as long as no feature is a linear combination of the others. That's why cell [24] can call np.linalg.inv(XTX).


